x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -58,6 +65,48 @@ func (l *loggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 	return resp, err
 }
 
+// heartBeat sends a ping message to the websocket server at regular intervals
+func (k *keepAlive) heartBeat(ctx context.Context, conn *websocket.Conn, start time.Time) context.CancelFunc {
+	ctx, cancel := context.WithCancel(ctx)
+	// set initial read deadline
+	conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	// set pong handler to update read deadline
+	conn.SetPongHandler(func(string) error {
+		k.log.Debugw("received pong message from websocket server")
+		k.metrics.pongMessageReceivedTime.Update(time.Since(start).Nanoseconds())
+		return conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	})
+
+	// set heartbeat ping routine
+	go func() {
+		ticker := time.NewTicker(k.cfg.KeepAlive.Interval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case now := <-ticker.C:
+				if conn != nil {
+					err := conn.WriteControl(websocket.PingMessage, nil, now.Add(k.cfg.KeepAlive.WriteControlDeadline))
+					if err != nil {
+						k.log.Debugw("error sending ping control frame to websocket server:", err)
+						k.metrics.writeControlErrors.Inc()
+						k.metrics.errorsTotal.Inc()
```

Comment:
Would it be the correct option to increase the total errors here ? This could increase error metrics bloat due to various network issues but according to design principle we currently have, this seems like the correct thing to do.

Reply:
Maybe add a ping errors metric? That would be possible to then be used to calculate the non-ping errors from the total.

Reply:
Yes already added it as write_control errors

Reply:
In that case, I think this is fine.